File: backend/app/models/task_model.py
from sqlalchemy import (
    Column,
    Integer,
    String,
    Text,
    DateTime,
    ForeignKey
)

from sqlalchemy.orm import relationship

from app.database import Base


class Task(Base):

    __tablename__ = "tasks"

    id = Column(Integer, primary_key=True, index=True)

    title = Column(String, nullable=False)

    description = Column(Text)

    due_date = Column(DateTime)

    user_id = Column(Integer, ForeignKey("users.id"))

    subject_id = Column(Integer, ForeignKey("subjects.id"))

    priority_id = Column(Integer, ForeignKey("priorities.id"))

    status_id = Column(Integer, ForeignKey("task_status.id"))

    category_id = Column(Integer, ForeignKey("categories.id"))

    user = relationship("User", back_populates="tasks")

    subject = relationship("Subject", back_populates="tasks")

    priority = relationship("Priority", back_populates="tasks")

    status = relationship("TaskStatus", back_populates="tasks")

    category = relationship("Category", back_populates="tasks")

    history = relationship("TaskHistory", back_populates="task")

File: frontend/lib/models/task_model.dart
class TaskModel {

  final int id;

  final String title;

  final String description;

  final String dueDate;

  final int subjectId;

  final int categoryId;

  final int priorityId;

  final int statusId;

  final String subjectName;

  final String categoryName;

  final String priorityName;

  final String statusName;

  TaskModel({

    required this.id,

    required this.title,

    required this.description,

    required this.dueDate,

    required this.subjectId,

    required this.categoryId,

    required this.priorityId,

    required this.statusId,

    required this.subjectName,

    required this.categoryName,

    required this.priorityName,

    required this.statusName,
  });

  factory TaskModel.fromJson(
      Map<String, dynamic> json) {

    return TaskModel(

      id: json['id'] ?? 0,

      title: json['title'] ?? '',

      description:
      json['description'] ?? '',

      dueDate:
      json['due_date'] ?? '',

      subjectId:
      json['subject_id'] ?? 0,

      categoryId:
      json['category_id'] ?? 0,

      priorityId:
      json['priority_id'] ?? 0,

      statusId:
      json['status_id'] ?? 0,

      subjectName:
      json['subject']['name'] ?? '',

      categoryName:
      json['category']['name'] ?? '',

      priorityName:
      json['priority']['name'] ?? '',

      statusName:
      json['status']['name'] ?? '',
    );
  }
}